returnn/hdf.py

```diff
@@ -346,6 +360,274 @@ def run(self):
+        self.target_sampling_rate = target_sampling_rate
+        self.skip_on_error = skip_on_error
+
+        self.out_hdfs = [self.output_path(f"{i + 1:0d}.hdf") for i in range(len(splits))]
+
+        num_cores = min(len(splits) + 1, 16)
+        mem_per_core = 1
+        self.rqmt = {"cpu": num_cores, "mem": num_cores * mem_per_core + 2, "time": 48}
+
+    def tasks(self):
+        yield Task("run", rqmt=self.rqmt, args=range(self.concurrent))
+
+    def run(self, index: int):
+        out_hdfs = list(util.chunks(self.out_hdfs, self.concurrent))[index]
+        splits = list(util.chunks(self.splits, self.concurrent))[index]
+        assert len(out_hdfs) == len(splits)
+        _logging.info(f"Writing audio to {out_hdfs} from {splits}.")
+
+        SimpleHDFWriter = get_returnn_simple_hdf_writer(self.returnn_root.get_path())
+        hdf_writers = [SimpleHDFWriter(filename=out_hdf.get_path(), dim=1) for out_hdf in out_hdfs]
+        segment_whitelists = []
+        for split in splits:
+            with uopen(split, "rt") as f:
+                segments_whitelist = {line.strip() for line in f if len(line.strip()) > 0}
+            segment_whitelists.append(segments_whitelist)
+
+        assert len(segment_whitelists) == len(out_hdfs)
+        all_whitelists = set.union(*segment_whitelists)
+
+        c = corpus.Corpus()
+        c.load(self.bliss_corpus.get_path())
+
+        pool = multiprocessing.Pool(max(self.rqmt["cpu"] - 1, 1))
+        apply_func = partial(
+            self._process_seq_handle_errors,
+            compress_format=self.compress_format,
+            compress_factor=self.compress_factor,
+            multi_channel_strategy=self.multi_channel_strategy,
+            output_dtype=self.output_dtype,
+            rounding_strategy=self.rounding,
+            round_factor=self.round_factor,
+            skip_on_error=self.skip_on_error,
+            target_sampling_rate=self.target_sampling_rate,
+        )
+        recs = (
+            # We send every segment of the recording grouped together with the audio to
+            # avoid having to read the audio file more than once.
+            #
+            # We only pass down the required audio/segment metadata instead of the whole
+            # `i6_core.lib.corpus.Segment` object to avoid excessive pickling overhead,
+            # stalling the worker processes (if we don't).
+            (rec.audio, segments)
+            for rec in c.all_recordings()
+            if len(
+                segments := [
+                    (full_name, segment.start, segment.end)
+                    for segment in rec.segments
+                    if (full_name := segment.fullname()) in all_whitelists
+                ]
+            )
+            > 0
+        )
+        # use imap instead of imap_unordered to have reproducible results, even though this
+        # causes a slight speed penalty
+        for results in pool.imap(apply_func, recs, chunksize=8):
```

Comment:
@albertz Since the users need to use the split files to recover the file ordering anyway, WDYT about switching this to `imap_unordered` for more speed? This will result in the HDFs having their segments ordered nondeterministically though, so two instances of the same job may not produce exactly the same HDFs, just "equivalent" HDFs.

Reply:
It should be deterministic.